delta/app/src/main/scala/ch/epfl/bluebrain/nexus/delta/serialization/EventSerializer.scala

```diff
@@ -0,0 +1,112 @@
+  }
+
+  override def toBinary(o: AnyRef): Array[Byte] = o match {
+    case e: PermissionsEvent  => e.asJson.noSpaces.getBytes(StandardCharsets.UTF_8)
+    case e: AclEvent          => e.asJson.noSpaces.getBytes(StandardCharsets.UTF_8)
+    case e: RealmEvent        => e.asJson.noSpaces.getBytes(StandardCharsets.UTF_8)
+    case e: OrganizationEvent => e.asJson.noSpaces.getBytes(StandardCharsets.UTF_8)
+    case e: ProjectEvent      => e.asJson.noSpaces.getBytes(StandardCharsets.UTF_8)
+    case _                    => throw new IllegalArgumentException(s"Unknown event type '${o.getClass.getCanonicalName}'")
+  }
+
+  override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = manifest match {
+    case `permissionsEventManifest`  => parseAndDecode[PermissionsEvent](bytes, manifest)
+    case `aclEventManifest`          => parseAndDecode[AclEvent](bytes, manifest)
+    case `realmEventManifest`        => parseAndDecode[RealmEvent](bytes, manifest)
+    case `organizationEventManifest` => parseAndDecode[OrganizationEvent](bytes, manifest)
+    case `projectEventManifest`      => parseAndDecode[ProjectEvent](bytes, manifest)
+    case _                           => throw new IllegalArgumentException(s"Unknown manifest '$manifest'")
+  }
+
+  private def parseAndDecode[E <: Event: Decoder](bytes: Array[Byte], manifest: String): E = {
+    val string = new String(bytes, StandardCharsets.UTF_8)
+    decode[E](string)
+      .getOrElse(throw new IllegalArgumentException(s"Unable to decode for manifest '$manifest' event '$string'"))
+  }
+}
+
+@nowarn("cat=unused")
+object EventSerializer {
+
+  final val permissionsEventManifest: String  = "permissions"
+  final val aclEventManifest: String          = "acl"
+  final val realmEventManifest: String        = "realm"
+  final val organizationEventManifest: String = "organization"
+  final val projectEventManifest: String      = "project"
+
+  implicit final private val configuration: Configuration =
+    Configuration.default.withStrictDecoding.withDiscriminator(keywords.tpe)
+
+  implicit final private val subjectCodec: Codec.AsObject[Subject]   = deriveConfiguredCodec[Subject]
+  implicit final private val identityCodec: Codec.AsObject[Identity] = deriveConfiguredCodec[Identity]
+
+  implicit final private val aclAddressEncoder: Encoder[AclAddress] = Encoder.encodeString.contramap(_.string)
+  implicit final private val aclAddressDecoder: Decoder[AclAddress] = Decoder.decodeString.emap { str =>
+    AclAddress.fromString(str).leftMap(_.getMessage)
+  }
+
+  final private case class AclEntry(identity: Identity, permissions: Set[Permission])
+  final private val aclEntryCodec                     = deriveConfiguredCodec[AclEntry]
+  implicit final private val aclEncoder: Encoder[Acl] =
+    Encoder.encodeList(aclEntryCodec).contramap(acl => acl.value.toList.map { case (id, perms) => AclEntry(id, perms) })
```

Comment:
Encoding it as a map is a problem ?

Reply:
you'd need a KeyEncoder / KeyDecoder for that, which yields the same amount of code and makes the json format different than how we serialize the events in the API. It's not worth it.